feat: Dynamic user count

mattvanvoorst-contentful opened the pull request:
- Channged the Discord's user count on the homepage. We now fetch the `approximate_member_count` from the Discord API and replace the default value of 200000. The user count is now also formatted by Intl.Numberformat, and rounded down to thousands.

All of this happens on getStaticProps of the index.js page, with a revalidation time of 60 seconds

mattmattvanvoorst committed:
feat: Dynamic user count
- Channged the Discord's user count on the homepage. We now fetch the `approximate_member_count` from the Discord API and replace the default value of 170000. The user count is now also formatted by Intl.Numberformat, and rounded down to thousands.

nickserv commented on src/pages/index.js:
```diff
@@ -31,15 +31,19 @@ const P = styled.p`
+        We’re a chat community of{" "}
+        {`${new Intl.NumberFormat("en-US").format(
+          Math.floor(userCount / 1000) * 1000,
+        )}+`}{" "}
```
I think it's better to use the default locale (for international users) with the exact user count (since we're getting an account count from the API anyway)
```suggestion
        {new Intl.NumberFormat().format(userCount)}{" "}
```

mattmattvanvoorst replied:
I'd stick with the `en-US` locale as it correctly formats separators and decimals. Since all the copy is in hardcoded english with US spelling I'd not want the usercount to be formatted according to a different locale (German for example switches commas and colons). 

I can for sure use the exact count, I tried to stay in line with the hardcoded value which was an estimate rounded down to thousands.

nickserv replied:
Good point, we can keep the locale consistent
```suggestion
        {new Intl.NumberFormat("en-US").format(userCount)}{" "}
```

mattvanvoorst-contentful committed:
Update src/pages/index.js

Render exact user count instead of rounded down on thousands

Co-authored-by: Nick McCurdy <nick@nickmccurdy.com>

mattmattvanvoorst committed:
feat: Dynamic user count
- Prettier

nickserv reviewed (approved):
Code looks good to me now 👍

nickserv commented:
Thinking before we deploy this: Will the count only update when we deploy? If so, is there an easy way around this that still involves a cache?

mattmattvanvoorst committed:
feat: Dynamic user count
- Added missing revalidate prop

mattvanvoorst-contentful commented:
That's a good shout, I just noticed that we're statically exporting this page. The solution would be to consider not using `next build` but `next start` so we can ISR the page.

vcarl commented:
Yeah we're currently hosted on Netlify, which requires static hosting. Modifying that to be hosted on a persistent server is an option, but would add enough complexity and additional cost (we're hosting for free on Netlify) that we should only do that once we want to offer more advanced functionality as well.

mattvanvoorst-contentful commented:
Doesn't Netlify nowadays offer ISR out of the box? I can't find any mention of it incurring any additional costs https://www.netlify.com/blog/2021/12/14/next.js-on-netlify-latest-update-introduces-complete-support-for-isr-now-with-revalidate/

louis-young commented on src/pages/index.js:
```diff
@@ -98,4 +99,30 @@ const Index = () => {
+  const fallbackUserCount = 200000;
+
+  try {
+    const r = await fetch(
```
I should be less lazy and check this myself but does this Fetch implementation throw on 4/5xx responses? I know the browser Fetch API only throws on network errors. 

I can see that it'd default if a situation occurred where it didn't throw and the `approximate_member_count` was still falsy but I thought I'd ask.

mattvanvoorst-contentful replied:
Yep! Anything that isn't a 200 will throw to the catch block

louis-young replied:
Awesome, thanks!

louis-young commented on src/pages/index.js:
```diff
@@ -31,15 +31,16 @@ const P = styled.p`
       <p>
-        We’re a chat community of 170,000+ React&nbsp;JS&nbsp;
+        We’re a chat community of{" "}
+        {new Intl.NumberFormat("en-US").format(userCount)} React&nbsp;JS&nbsp;
```
Are we intentionally mixing non-breaking spaces and JSX expressions here? I know I'm a massive pedant but would it be more consistent to use a non-breaking space instead of the JSX expression here?

louis-young replied:
Also, is there a reason you prefer `Intl.NumberFormat` over `Number.toLocaleString`? I'm genuinely curious, not suggesting a change.

mattvanvoorst-contentful replied:
The mix of whitespace syntax is part prettier and part old copy.

No preference; Intl is the first thing that came to mind to format the number

louis-young replied:
Fair enough, looks good to me either way 👍

nickserv replied:
I believe the non-breaking spaces were intentionally placed here to force "React JS ⚛️" to appear on one line, without forcing breaks around the user count.

louis-young replied:
Ah, that'd make sense!

louis-young commented on src/pages/index.js:
```diff
@@ -98,4 +99,30 @@ const Index = () => {
+    const data = await r.json();
+
+    return {
+      revalidate: 60,
```
Apologies for being a massive pedant again but is there any value in variablising the revalidate time?

mattvanvoorst-contentful replied:
This is the only page that uses ISR; there wouldn't by any benefit rn

louis-young replied:
I see, that makes sense 👍

louis-young reviewed (approved):
Looks great Matt, nice one 🎉 

Feel free to ignore any/all of my pedantic comments 😄

nickserv commented:
From what I understand, [Netlify supports both static and dynamic Next deployments](https://docs.netlify.com/integrations/frameworks/#next-js), though I'm not sure if they're cached at the edge so there may be an additional performance penalty. We're currently using static, but for this feature it would probably make sense to try making it dynamic.

@vcarl How would you feel about deploying this? Our free plan includes 125,000 requests and 100 hours of runtime per month, which we're not using yet.

nickserv committed:
Merge branch 'main' into feature/dynamic-user-count

nickserv commented on src/pages/index.js:
```diff
@@ -98,4 +99,30 @@ const Index = () => {
 };
 
+export const getStaticProps = async () => {
+  const fallbackUserCount = 200000;
```
Minor nitpick: Would probably be easier to read with separators (assuming it's supported)
```suggestion
  const fallbackUserCount = 200_000;
```

mattvanvoorst-contentful replied:
Good shout; updated that. 

I also changed the build command to `yarn start` instead of `yarn build` in the netlify.toml file, that way we can test the ISR on this PR

mattmattvanvoorst committed:
feat: Dynamic user count
- Made the `fallbackUserCount` better legible
- Changed the CICD command for Netlify from export & build to start, so we can use ISR rendering

mattmattvanvoorst committed:
Merge remote-tracking branch 'origin/feature/dynamic-user-count' into feature/dynamic-user-count

mattmattvanvoorst committed:
feat: Dynamic user count
- Updated build command

mattmattvanvoorst committed:
feat: Dynamic user count
- Updated build command, once again. Start seems not needed

mattmattvanvoorst committed:
feat: Dynamic user count
- Updated build command

mattmattvanvoorst committed:
feat: Dynamic user count
- Updated build command

nickserv committed:
Update publish directory for updated build command

nickserv commented:
Reposting @mattmattvanvoorst's comment so it can be read after I resolve the review:
> I also changed the build command to `yarn start` instead of `yarn build` in the netlify.toml file, that way we can test the ISR on this PR

nickserv committed:
Move `next build` from Netlify to `build` script

nickserv commented:
The builds are now passing, but the site 404s.

nickserv committed:
Enable Next.js Runtime
